### PR TITLE
Manifest handling.

### PR DESCRIPTION
**Pull request by twilly**
A manifest is a file that lists files and their hash digests. The canonical form is `<hash type>:<digest>[sp][sp]<path>\n` lexicographically ordered by path.

This package implements encoding manifests and accessing their content by path or digest. Manifests can be created from a bucket where the bucket is walked and a manifest is created for every file contained within.

NOTE: This is a rebased PR of #1499 to capture required CI changes. See that PR for previous discussion.

### CODE REVIEW

**Review by pkwarren — APPROVED** [review]

**Review by unmultimedio — APPROVED** [review]

File: private/pkg/manifest/manifest.go [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+// ManifestError occurs when a manifest is malformed.
+type ManifestError struct {
+	lineno  int
+	msg     string
+	wrapped error
+}
+
+var _ error = (*ManifestError)(nil)
+
+func newManifestError(lineno int, msg string) *ManifestError {
```
Comment (unmultimedio) [inline private/pkg/manifest/manifest.go:105]:
```suggestion
func newError(lineno int, msg string) *ManifestError {
```

File: private/pkg/manifest/manifest.go [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+func (d *Digest) Bytes() []byte {
+	return d.digest
+}
+
+func (d *Digest) Hex() string {
+	return d.hexstr
+}
+
+// ManifestError occurs when a manifest is malformed.
+type ManifestError struct {
```
Comment (unmultimedio) [inline private/pkg/manifest/manifest.go:97]:
```suggestion
type Error struct {
```

File: private/pkg/manifest/manifest.go [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+var _ error = (*ManifestError)(nil)
+
+func newManifestError(lineno int, msg string) *ManifestError {
+	return &ManifestError{
+		lineno: lineno,
+		msg:    msg,
+	}
+}
+func newManifestErrorFromError(lineno int, err error) *ManifestError {
```
Comment (unmultimedio) [inline private/pkg/manifest/manifest.go:111]:
```suggestion
func newError(lineno int, err error) *Error {
```

File: private/pkg/manifest/manifest.go [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+	pathToDigest  map[string]Digest
+	digestToPaths map[string][]string
+	hash          sha3.ShakeHash
+}
+
+var _ encoding.TextMarshaler = (*Manifest)(nil)
+var _ encoding.TextUnmarshaler = (*Manifest)(nil)
+
+// NewManifest creates an empty manifest.
+func NewManifest() *Manifest {
```
Comment (unmultimedio) [inline private/pkg/manifest/manifest.go:138]:
```suggestion
func New() *Manifest {
```

File: private/pkg/manifest/manifest.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+	return &Manifest{
+		pathToDigest:  make(map[string]Digest),
+		digestToPaths: make(map[string][]string),
+		hash:          sha3.NewShake256(),
+	}
+}
+
+// NewManifestFromReader builds a manifest an encoded manifest reader.
+func NewManifestFromReader(manifest io.Reader) (*Manifest, error) {
```
Comment (unmultimedio) [inline private/pkg/manifest/manifest.go:147]:
```suggestion
func NewFromReader(manifest io.Reader) (*Manifest, error) {
```

File: private/pkg/manifest/manifest.go [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return m, nil
+}
+
+// NewManifestFromBucket creates a manifest from a bucket.
+func NewManifestFromBucket(
```
Comment (unmultimedio) [inline private/pkg/manifest/manifest.go:178]:
```suggestion
func NewFromBucket(
```

**Review by unmultimedio — COMMENTED** [review]
A few suggestions removing stutter I just realized after using it in the backend (pkg name is already `manifest`). This can be addressed since this is not in `main` yet.

**Pull request merged**